// devextreme-vue/cjs/core/__tests__/selectbox.test.js
/*!
 * devextreme-vue
 * Version: 23.2.3
 * Build date: Tue Nov 28 2023
 *
 * Copyright (c) 2012 - 2023 Developer Express Inc. ALL RIGHTS RESERVED
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file in the root of the project for details.
 *
 * https://github.com/DevExpress/devextreme-vue
 */

"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var test_utils_1 = require("@vue/test-utils");
var vue_1 = require("vue");
var select_box_1 = __importDefault(require("../../select-box"));
var text_box_1 = __importDefault(require("../../text-box"));
jest.setTimeout(1000);
beforeEach(function () {
    jest.clearAllMocks();
});
describe('template rendering', function () {
    it('field template rendered', function () {
        var vm = vue_1.defineComponent({
            template: "<dx-select-box :data-source=\"dataSource\" field-template=\"field\" id=\"component\">\n                    <template #field=\"\">\n                        <dx-text-box value=\"text\" />\n                    </template>\n                </dx-select-box >",
            data: function () {
                return {
                    dataSource: [{ ID: 1 }],
                };
            },
            components: {
                DxSelectBox: select_box_1.default,
                DxTextBox: text_box_1.default,
            },
        });
        var wrapper = test_utils_1.mount(vm);
        expect(wrapper.getComponent('#component').vm.$el.children).toHaveLength(1);
    });
});
