// devextreme-vue/cjs/core/__tests__/templates-discovering.test.js
/*!
 * devextreme-vue
 * Version: 23.2.3
 * Build date: Tue Nov 28 2023
 *
 * Copyright (c) 2012 - 2023 Developer Express Inc. ALL RIGHTS RESERVED
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file in the root of the project for details.
 *
 * https://github.com/DevExpress/devextreme-vue
 */

"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var test_utils_1 = require("@vue/test-utils");
var vue_1 = require("vue");
var templates_discovering_1 = require("../templates-discovering");
describe('templates-discovering (vue 3)', function () {
    it('discovers named scoped slot', function () {
        var template = '<template #slot_name=\'_\'/>';
        expect(getDiscoveredTemplates(template)).toEqual(['slot_name']);
    });
    it('discovers named not-scoped slot', function () {
        var template = '<template #slot_name/>';
        expect(getDiscoveredTemplates(template)).toEqual(['slot_name']);
    });
    // Vue doesn't recognize this as slot
    it('doesn\'t discover not-scoped not-named slot', function () {
        var template = '<template/>';
        expect(getDiscoveredTemplates(template)).toEqual([]);
    });
    // to avoid creating templates from config-components
    it('doesn\'t discover implicit default slot', function () {
        var template = '<div>abc</div>';
        expect(getDiscoveredTemplates(template)).toEqual([]);
    });
    // to avoid creating templates from config-components
    it('doesn\'t discover custom component', function () {
        var template = '<customComponent #default=\'_\'/>';
        expect(getDiscoveredTemplates(template)).toEqual([]);
    });
});
function getDiscoveredTemplates(template) {
    var actual;
    var vm = vue_1.defineComponent({
        template: "<container>" + template + "</container>",
        components: {
            container: {
                render: function () {
                    actual = templates_discovering_1.discover(this);
                    return null;
                },
            },
            customComponent: {
                render: function () {
                    return null;
                },
            },
        },
    });
    test_utils_1.mount(vm);
    return Object.keys(actual);
}
